File: sdk/javascript/examples.js
/**
 * Example usage of RAG JavaScript SDK
 * 
 * This file demonstrates various use cases for the RAG JavaScript SDK.
 * 
 * Usage:
 *   node examples.js
 * 
 * Or in browser:
 *   <script src="rag-sdk.js"></script>
 *   <script src="examples.js"></script>
 */

// For Node.js, import the SDK
// const { RAGClient } = require('./rag-sdk.js');

// Examples
const examples = {
  
  /**
   * Example 1: Basic usage with login
   */
  async basicUsage() {
    console.log('\n=== Example 1: Basic Usage ===');
    
    const client = new RAGClient({
      baseUrl: 'http://localhost:8000',
      tenantId: 'demo-tenant'
    });
    
    try {
      // Login
      const token = await client.login('user@example.com', 'password');
      console.log(`✓ Logged in. Token expires in ${token.expiresIn} seconds`);
      
      // Health check
      const health = await client.healthCheck();
      console.log(`✓ API Health: ${health.status}`);
      
    } catch (error) {
      if (error.name === 'AuthenticationError') {
        console.log(`✗ Authentication failed: ${error.message}`);
      } else {
        console.log(`✗ Error: ${error.message}`);
      }
    }
  },

  /**
   * Example 2: Using API key authentication
   */
  async apiKeyAuth() {
    console.log('\n=== Example 2: API Key Authentication ===');
    
    // No login needed with API key
    const client = new RAGClient({
      baseUrl: 'http://localhost:8000',
      tenantId: 'demo-tenant',
      apiKey: 'rag_your_api_key_here'
    });
    
    try {
      // Directly use search without login
      const results = await client.search('machine learning', { topK: 3 });
      console.log(`✓ Found ${results.length} results using API key`);
      
    } catch (error) {
      console.log(`✗ API error: ${error.message}`);
    }
  },

  /**
   * Example 3: RAG query with citations
   */
  async ragQuery() {
    console.log('\n=== Example 3: RAG Query ===');
    
    const client = new RAGClient({
      baseUrl: 'http://localhost:8000',
      tenantId: 'demo-tenant',
      apiKey: 'rag_your_api_key_here'
    });
    
    try {
      const response = await client.ragQuery('What is machine learning?', {
        topK: 5,
        sessionId: 'session-001'
      });
      
      console.log(`✓ Answer: ${response.answer}`);
      console.log(`  Confidence: ${response.confidence.toFixed(2)}`);
      console.log(`  Citations: ${response.citations.length}`);
      
      response.citations.slice(0, 3).forEach((citation, i) => {
        console.log(`    ${i + 1}. ${citation.doc_id} (source: ${citation.source})`);
      });
      
    } catch (error) {
      console.log(`✗ API error: ${error.message}`);
    }
  },

  /**
   * Example 4: Structured data extraction
   */
  async dataExtraction() {
    console.log('\n=== Example 4: Data Extraction ===');
    
    const client = new RAGClient({
      baseUrl: 'http://localhost:8000',
      tenantId: 'demo-tenant',
      apiKey: 'rag_your_api_key_here'
    });
    
    // Define extraction schema
    const schema = {
      type: 'object',
      properties: {
        company_name: {
          type: 'string',
          description: 'Name of the company'
        },
        founded_year: {
          type: 'integer',
          description: 'Year the company was founded'
        },
        headquarters: {
          type: 'string',
          description: 'Location of headquarters'
        },
        employees: {
          type: 'integer',
          description: 'Number of employees'
        }
      },
      required: ['company_name']
    };
    
    try {
      const result = await client.extract(
        'Extract company information from the about page',
        schema,
        { topK: 5, minConfidence: 0.7 }
      );
      
      if (result.success) {
        console.log('✓ Extraction successful');
        console.log(`  Confidence: ${result.confidence.toFixed(2)}`);
        console.log(`  Data: ${JSON.stringify(result.data, null, 2)}`);
      } else {
        console.log('✗ Extraction failed');
        console.log(`  Errors: ${result.validationErrors.join(', ')}`);
      }
      
    } catch (error) {
      console.log(`✗ API error: ${error.message}`);
    }
  },

  /**
   * Example 5: Async extraction with job tracking
   */
  async asyncExtraction() {
    console.log('\n=== Example 5: Async Extraction ===');
    
    const client = new RAGClient({
      baseUrl: 'http://localhost:8000',
      tenantId: 'demo-tenant',
      apiKey: 'rag_your_api_key_here'
    });
    
    const schema = {
      type: 'object',
      properties: {
        employees: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              name: { type: 'string' },
              email: { type: 'string' },
              department: { type: 'string' }
            }
          }
        }
      }
    };
    
    try {
      console.log('Starting async extraction...');
      const result = await client.extractWithJob(
        'Extract all employee information',
        schema,
        {
          schemaName: 'employee_list',
          topK: 10,
          pollInterval: 2000,
          maxWait: 60000
        }
      );
      
      if (result.success) {
        console.log('✓ Async extraction completed');
        console.log(`  Job ID: ${result.jobId}`);
        console.log(`  Found ${result.data.employees?.length || 0} employees`);
      } else {
        console.log('✗ Async extraction failed');
        console.log(`  Errors: ${result.validationErrors.join(', ')}`);
      }
      
    } catch (error) {
      console.log(`✗ API error: ${error.message}`);
    }
  },

  /**
   * Example 6: Error handling
   */
  async errorHandling() {
    console.log('\n=== Example 6: Error Handling ===');
    
    const client = new RAGClient({
      baseUrl: 'http://localhost:8000',
      tenantId: 'demo-tenant'
    });
    
    try {
      // Try to search without authentication
      await client.search('test');
      
    } catch (error) {
      if (error.name === 'AuthenticationError') {
        console.log(`✓ Caught authentication error: ${error.message}`);
      } else if (error.name === 'RateLimitError') {
        console.log(`✓ Caught rate limit error: ${error.message}`);
        console.log(`  Retry after: ${error.retryAfter} seconds`);
      } else if (error.name === 'RAGError') {
        console.log(`✓ Caught API error: ${error.message}`);
        console.log(`  Status code: ${error.statusCode}`);
      } else {
        console.log(`✗ Unexpected error: ${error.message}`);
      }
    }
  },

  /**
   * Example 7: Multi-turn conversation with session
   */
  async conversationSession() {
    console.log('\n=== Example 7: Conversation Session ===');
    
    const sessionId = `session-${Date.now()}`;
    console.log(`Session ID: ${sessionId}`);
    
    const client = new RAGClient({
      baseUrl: 'http://localhost:8000',
      tenantId: 'demo-tenant',
      apiKey: 'rag_your_api_key_here'
    });
    
    const questions = [
      'What is machine learning?',
      'What are the types of machine learning?',
      'Give me an example of supervised learning.'
    ];
    
    for (let i = 0; i < questions.length; i++) {
      try {
        const response = await client.ragQuery(questions[i], { sessionId });
        
        console.log(`\nQ${i + 1}: ${questions[i]}`);
        console.log(`A${i + 1}: ${response.answer.substring(0, 100)}...`);
        
      } catch (error) {
        console.log(`✗ Error in turn ${i + 1}: ${error.message}`);
      }
    }
  },

  /**
   * Example 8: Batch search operations
   */
  async batchSearch() {
    console.log('\n=== Example 8: Batch Search ===');
    
    const queries = [
      'machine learning',
      'deep learning',
      'neural networks',
      'natural language processing'
    ];
    
    const client = new RAGClient({
      baseUrl: 'http://localhost:8000',
      tenantId: 'demo-tenant',
      apiKey: 'rag_your_api_key_here'
    });
    
    for (const query of queries) {
      try {
        const results = await client.search(query, { topK: 3 });
        console.log(`✓ '${query}': ${results.length} results`);
        
      } catch (error) {
        console.log(`✗ '${query}': Error - ${error.message}`);
      }
    }
  },

  /**
   * Example 9: Quick convenience functions
   */
  async quickFunctions() {
    console.log('\n=== Example 9: Quick Functions ===');
    
    // Note: These functions require the SDK to be imported
    // const { search, ragQuery } = require('./rag-sdk.js');
    
    const config = {
      baseUrl: 'http://localhost:8000',
      tenantId: 'demo-tenant',
      apiKey: 'rag_your_api_key_here'
    };
    
    try {
      // Quick search
      // const results = await search(config, 'machine learning', { topK: 5 });
      console.log('✓ Quick search: (uncomment to run)');
      
      // Quick RAG query
      // const answer = await ragQuery(config, 'What is AI?');
      console.log('✓ Quick RAG: (uncomment to run)');
      
    } catch (error) {
      console.log(`✗ Error: ${error.message}`);
    }
  },

  /**
   * Example 10: Using with async/await and Promise.all
   */
  async parallelRequests() {
    console.log('\n=== Example 10: Parallel Requests ===');
    
    const client = new RAGClient({
      baseUrl: 'http://localhost:8000',
      tenantId: 'demo-tenant',
      apiKey: 'rag_your_api_key_here'
    });
    
    const queries = [
      'machine learning basics',
      'deep learning introduction',
      'neural networks explained'
    ];
    
    try {
      console.log('Running parallel searches...');
      
      // Execute searches in parallel
      const promises = queries.map(query => 
        client.search(query, { topK: 3 })
          .then(results => ({ query, results, success: true }))
          .catch(error => ({ query, error: error.message, success: false }))
      );
      
      const results = await Promise.all(promises);
      
      results.forEach(({ query, results, error, success }) => {
        if (success) {
          console.log(`✓ '${query}': ${results.length} results`);
        } else {
          console.log(`✗ '${query}': ${error}`);
        }
      });
      
    } catch (error) {
      console.log(`✗ Error: ${error.message}`);
    }
  }

};


// Run examples
async function runExamples() {
  console.log('=' .repeat(70));
  console.log('RAG SDK JavaScript Examples');
  console.log('=' .repeat(70));
  
  const exampleList = [
    examples.basicUsage,
    examples.apiKeyAuth,
    examples.ragQuery,
    examples.dataExtraction,
    examples.asyncExtraction,
    examples.errorHandling,
    examples.conversationSession,
    examples.batchSearch,
    examples.quickFunctions,
    examples.parallelRequests
  ];
  
  for (const example of exampleList) {
    try {
      await example();
    } catch (error) {
      console.log(`\n✗ Example failed: ${error.message}`);
    }
  }
  
  console.log('\n' + '='.repeat(70));
  console.log('Examples completed!');
  console.log('='.repeat(70));
}

// Run if in Node.js
if (typeof module !== 'undefined' && module.exports) {
  // Export for use as module
  module.exports = examples;
  
  // Run examples if executed directly
  if (require.main === module) {
    runExamples();
  }
}

File: sdk/python/examples.py
"""
Example usage of RAG Python SDK

This file demonstrates various use cases for the RAG SDK.
"""

from rag_sdk import RAGClient, AuthenticationError, RateLimitError, APIError


def example_basic_usage():
    """Example 1: Basic usage with login."""
    print("\n=== Example 1: Basic Usage ===")

    client = RAGClient(base_url="http://localhost:8000", tenant_id="demo-tenant")

    try:
        # Login
        token = client.login("user@example.com", "password")
        print(f"✓ Logged in. Token expires in {token.expires_in} seconds")

        # Health check
        health = client.health_check()
        print(f"✓ API Health: {health['status']}")

    except AuthenticationError as e:
        print(f"✗ Authentication failed: {e}")
    finally:
        client.close()


def example_api_key_auth():
    """Example 2: Using API key authentication."""
    print("\n=== Example 2: API Key Authentication ===")

    # No login needed with API key
    client = RAGClient(
        base_url="http://localhost:8000",
        tenant_id="demo-tenant",
        api_key="rag_your_api_key_here",
    )

    try:
        # Directly use search without login
        results = client.search("machine learning", top_k=3)
        print(f"✓ Found {len(results)} results using API key")

    except APIError as e:
        print(f"✗ API error: {e}")
    finally:
        client.close()


def example_context_manager():
    """Example 3: Using context manager."""
    print("\n=== Example 3: Context Manager ===")

    try:
        with RAGClient(
            base_url="http://localhost:8000",
            tenant_id="demo-tenant",
            api_key="rag_your_api_key_here",
        ) as client:
            results = client.search("deep learning", top_k=5)
            print(f"✓ Found {len(results)} documents")

            for i, result in enumerate(results[:3], 1):
                print(f"  {i}. {result.doc_id} (score: {result.score:.2f})")

    except Exception as e:
        print(f"✗ Error: {e}")


def example_rag_query():
    """Example 4: RAG query with citations."""
    print("\n=== Example 4: RAG Query ===")

    with RAGClient(
        base_url="http://localhost:8000",
        tenant_id="demo-tenant",
        api_key="rag_your_api_key_here",
    ) as client:
        try:
            response = client.rag_query(
                "What is machine learning?", top_k=5, session_id="session-001"
            )

            print(f"✓ Answer: {response.answer}")
            print(f"  Confidence: {response.confidence:.2f}")
            print(f"  Citations: {len(response.citations)}")

            for citation in response.citations[:3]:
                print(
                    f"    - {citation.get('doc_id')} (source: {citation.get('source')})"
                )

        except APIError as e:
            print(f"✗ API error: {e}")


def example_data_extraction():
    """Example 5: Structured data extraction."""
    print("\n=== Example 5: Data Extraction ===")

    with RAGClient(
        base_url="http://localhost:8000",
        tenant_id="demo-tenant",
        api_key="rag_your_api_key_here",
    ) as client:
        # Define extraction schema
        schema = {
            "type": "object",
            "properties": {
                "company_name": {
                    "type": "string",
                    "description": "Name of the company",
                },
                "founded_year": {
                    "type": "integer",
                    "description": "Year the company was founded",
                },
                "headquarters": {
                    "type": "string",
                    "description": "Location of headquarters",
                },
                "employees": {"type": "integer", "description": "Number of employees"},
            },
            "required": ["company_name"],
        }

        try:
            result = client.extract(
                "Extract company information from the about page",
                schema=schema,
                top_k=5,
                min_confidence=0.7,
            )

            if result.success:
                print(f"✓ Extraction successful")
                print(f"  Confidence: {result.confidence:.2f}")
                print(f"  Data: {result.data}")
            else:
                print(f"✗ Extraction failed")
                print(f"  Errors: {result.validation_errors}")

        except APIError as e:
            print(f"✗ API error: {e}")


def example_async_extraction():
    """Example 6: Async extraction with job tracking."""
    print("\n=== Example 6: Async Extraction ===")

    with RAGClient(
        base_url="http://localhost:8000",
        tenant_id="demo-tenant",
        api_key="rag_your_api_key_here",
    ) as client:
        schema = {
            "type": "object",
            "properties": {
                "employees": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "name": {"type": "string"},
                            "email": {"type": "string"},
                            "department": {"type": "string"},
                        },
                    },
                }
            },
        }

        try:
            result = client.extract_with_job(
                "Extract all employee information",
                schema=schema,
                schema_name="employee_list",
                top_k=10,
                poll_interval=2,
                max_wait=60,
            )

            if result.success:
                print(f"✓ Async extraction completed")
                print(f"  Job ID: {result.job_id}")
                print(f"  Found {len(result.data.get('employees', []))} employees")
            else:
                print(f"✗ Async extraction failed")
                print(f"  Errors: {result.validation_errors}")

        except APIError as e:
            print(f"✗ API error: {e}")


def example_error_handling():
    """Example 7: Error handling."""
    print("\n=== Example 7: Error Handling ===")

    with RAGClient(base_url="http://localhost:8000", tenant_id="demo-tenant") as client:
        try:
            # Try to search without authentication
            results = client.search("test")

        except AuthenticationError as e:
            print(f"✓ Caught authentication error: {e}")

        except RateLimitError as e:
            print(f"✓ Caught rate limit error: {e}")
            print(f"  Retry after: {e.retry_after} seconds")

        except APIError as e:
            print(f"✓ Caught API error: {e}")
            print(f"  Status code: {e.status_code}")

        except Exception as e:
            print(f"✗ Unexpected error: {e}")


def example_conversation_session():
    """Example 8: Multi-turn conversation with session."""
    print("\n=== Example 8: Conversation Session ===")

    import uuid

    session_id = str(uuid.uuid4())
    print(f"Session ID: {session_id}")

    with RAGClient(
        base_url="http://localhost:8000",
        tenant_id="demo-tenant",
        api_key="rag_your_api_key_here",
    ) as client:
        questions = [
            "What is machine learning?",
            "What are the types of machine learning?",
            "Give me an example of supervised learning.",
        ]

        for i, question in enumerate(questions, 1):
            try:
                response = client.rag_query(question, session_id=session_id)

                print(f"\nQ{i}: {question}")
                print(f"A{i}: {response.answer[:100]}...")

            except APIError as e:
                print(f"✗ Error in turn {i}: {e}")


def example_batch_search():
    """Example 9: Batch search operations."""
    print("\n=== Example 9: Batch Search ===")

    queries = [
        "machine learning",
        "deep learning",
        "neural networks",
        "natural language processing",
    ]

    with RAGClient(
        base_url="http://localhost:8000",
        tenant_id="demo-tenant",
        api_key="rag_your_api_key_here",
    ) as client:
        for query in queries:
            try:
                results = client.search(query, top_k=3)
                print(f"✓ '{query}': {len(results)} results")

            except APIError as e:
                print(f"✗ '{query}': Error - {e}")


def example_quick_functions():
    """Example 10: Quick convenience functions."""
    print("\n=== Example 10: Quick Functions ===")

    from rag_sdk import search, rag_query

    config = {
        "baseUrl": "http://localhost:8000",
        "tenantId": "demo-tenant",
        "apiKey": "rag_your_api_key_here",
    }

    try:
        # Quick search
        results = search(config, "machine learning", top_k=5)
        print(f"✓ Quick search: {len(results)} results")

        # Quick RAG query
        answer = rag_query(config, "What is AI?")
        print(f"✓ Quick RAG: {answer.answer[:50]}...")

    except Exception as e:
        print(f"✗ Error: {e}")


if __name__ == "__main__":
    print("=" * 70)
    print("RAG SDK Python Examples")
    print("=" * 70)

    # Run examples
    examples = [
        example_basic_usage,
        example_api_key_auth,
        example_context_manager,
        example_rag_query,
        example_data_extraction,
        example_async_extraction,
        example_error_handling,
        example_conversation_session,
        example_batch_search,
        example_quick_functions,
    ]

    for example in examples:
        try:
            example()
        except Exception as e:
            print(f"\n✗ Example failed: {e}")

    print("\n" + "=" * 70)
    print("Examples completed!")
    print("=" * 70)
